File: src/api/git/tools.ts
import { invoke } from '@tauri-apps/api/core';

export interface ToolCommands {
  diff: string | null;
  merge: string | null;
}

export interface ToolsConfig {
  default: ToolCommands;
  extensions: { [key: string]: ToolCommands };
}

export async function loadToolsConfig(): Promise<ToolsConfig> {
  return await invoke('load_tools_config');
}

export async function saveToolsConfig(config: ToolsConfig): Promise<void> {
  await invoke('save_tools_config', { config });
}

export async function detectInstalledTools(): Promise<ToolsConfig> {
  return await invoke('detect_installed_tools');
}

export async function launchExternalDiff(filePath: string): Promise<void> {
  await invoke('launch_external_diff', { filePath });
}

export async function launchExternalMerge(filePath: string): Promise<void> {
  await invoke('launch_external_merge', { filePath });
}

export async function testToolCommand(commandStr: string): Promise<void> {
  await invoke('test_tool_command', { commandStr });
}

File: src-tauri/src/commands/tools.rs
use crate::git;
use crate::state::AppState;
use tauri::State;

#[tauri::command]
pub fn load_tools_config() -> crate::error::Result<git::ToolsConfig> {
    git::load_tools_config()
}

#[tauri::command]
pub fn save_tools_config(config: git::ToolsConfig) -> crate::error::Result<()> {
    git::save_tools_config(&config)
}

#[tauri::command]
pub fn detect_installed_tools() -> crate::error::Result<git::ToolsConfig> {
    git::detect_installed_tools()
}

#[tauri::command]
pub fn launch_external_diff(
    file_path: String,
    state: State<'_, AppState>,
) -> crate::error::Result<()> {
    let repo_path = state.repo_path.lock().unwrap();
    if let Some(p) = repo_path.as_ref() {
        git::launch_external_diff(p, file_path)
    } else {
        Err(crate::error::GitError::CommandFailed(
            "No repository is currently open.".to_string(),
        ))
    }
}

#[tauri::command]
pub fn launch_external_merge(
    file_path: String,
    state: State<'_, AppState>,
) -> crate::error::Result<()> {
    let repo_path = state.repo_path.lock().unwrap();
    if let Some(p) = repo_path.as_ref() {
        git::launch_external_merge(p, file_path)
    } else {
        Err(crate::error::GitError::CommandFailed(
            "No repository is currently open.".to_string(),
        ))
    }
}

#[tauri::command]
pub fn test_tool_command(command_str: String) -> crate::error::Result<()> {
    git::test_tool_command(command_str)
}
